### PR TITLE
#1: InvoiceExpress API

### PR DESCRIPTION
**Pull request by BeeMargarida**
| - | - |
| --- | --- |
| Issue | https://github.com/ripe-tech/peri-invoicing/issues/1 |
| Dependencies | -- |
| Decisions | - Methods to create an invoice, change its status and get a PDF of an invoice. |
| Animated GIF | -- |

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,3 +1,21 @@
...
+## Configuration
+
+| Name             | Type  | Default | Description                                                   |
+| ---------------- | ----- | ------- | ------------------------------------------------------------- |
+| **ACCOUNT_NAME** | `str` | `None`  | The name of the Invoice Express account.                      |
+| **API_KEY**      | `str` | `None`  | The API key used to make requests to the Invoice Express API. |
+
+## License
+
+UPS API for Javascript is currently licensed under the [Apache License, Version 2.0](http://www.apache.org/licenses/).
```
Comment (joao-conde) [inline README.md]: Invoice Express API

File: js/base.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+export class API extends mix(BaseAPI).with(InvoiceAPI) {
+    constructor(kwargs = {}) {
+        super(kwargs);
+        this.accountName = conf("ACCOUNT_NAME", null);
+        this.apiKey = conf("API_KEY", null);
+        this.accountName = kwargs.accountName === undefined ? this.accountName : kwargs.accountName;
+        this.apiKey = kwargs.apiKey === undefined ? this.apiKey : kwargs.apiKey;
+
+        this.invoicingBaseUrl = `https://${this.accountName}.app.invoicexpress.com/`;
```
Comment (joao-conde) [inline js/base.js]: Allow override via `kwargs` and delete new line above.

File: js/base.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    static async load() {
+        await load();
+    }
+
+    async build(method, url, options = {}) {
+        await super.build(method, url, options);
+        options.headers = options.headers !== undefined ? options.headers : {};
+        options.kwargs = options.kwargs !== undefined ? options.kwargs : {};
+        options.params = options.params !== undefined ? options.params : {};
+
```
Comment (joao-conde) [inline js/base.js]: Think Joao will want this new line gone

File: js/invoice.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
+export const InvoiceAPI = superclass =>
+    class extends superclass {
+        async createInvoice(payload) {
```
Comment (joao-conde) [inline js/invoice.js]: Add `options` parameter that defaults to `{}` and pass it along to the request (similar to UPS API JS)

File: js/invoice.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    class extends superclass {
+        async createInvoice(payload) {
+            const url = this.invoicingBaseUrl + "invoices.json";
+            const response = await this.post(url, {
+                dataJ: payload
+            });
+            return response;
+        }
+
+        async changeInvoiceStatus(invoiceId, payload) {
```
Comment (joao-conde) [inline js/invoice.js]: Same as above

File: package.json [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        "lint": "eslint rollup.config.js \"./{js,test}/**/*.{js,json}\"",
+        "lint-fix": "eslint rollup.config.js \"./{js,test}/**/*.{js,json}\" --fix",
+        "prettier": "prettier rollup.config.js \"./*.{js,json}\" \"./{js,test}/**/*.{js,json}\" --write",
+        "pretty": "npm run prettier && npm run lint-fix",
+        "test": "mocha --recursive",
+        "upgrade": "npx sort-package-json && ncu -u",
+        "watch": "rollup -c -w"
+    },
+    "dependencies": {
+        "yonius": "^0.6.5"
```
Comment (joao-conde) [inline package.json]: If possible, let's start with upgraded dependencies maybe. At least for yonius (maybe others as well)

File: test/api/base.js [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+const invoiceExpress = require("../..");
+
+describe("API", function() {
+    it("should be able to instantiate the API", async () => {
+        const api = new invoiceExpress.API({
+            accountName: "name",
+            apiKey: "key"
+        });
+        assert.strictEqual(Boolean(api.accountName), true);
+        assert.strictEqual(Boolean(api.apiKey), true);
```
Comment (joao-conde) [inline test/api/base.js]: Maybe add one for `assert.strictEqual(Boolean(api.invoiceBaseUrl), true);`
Also, maybe test for `api.accountName === "name"` and same for key

**Review by joao-conde — CHANGES_REQUESTED** [review]

**Comment by BeeMargarida** [comment]
@joao-conde Just pinging to remind you about this

**Review by joao-conde — APPROVED** [review]

**Review by joamag — APPROVED** [review]

**Pull request merged**